### DISCUSSION TITLE
Use conventional commit messages

### DISCUSSION THREAD

**Discussion by 2bndy5** [cpp-linter/cpp-linter#122] [Ideas]
I want to better my git game with proper changelogs (preferably generated from `git log`). To do this, I need to follow a standard for formatting commit messages, specifically the commit title (not the messages' extra details in the message body).

# My research on commit message standards
The most popular standard I've come across is the [conventional commit](https://www.conventionalcommits.org) (which evolved from the [angular project's guidelines](https://github.com/angular/angular/blob/68a6a07/CONTRIBUTING.md#-commit-message-format)).

There are a bunch of [tools to enforce this](https://www.conventionalcommits.org/en/about/#tooling-for-conventional-commits). Ideally, I want
- [ ] a pre-commit hook that only runs on the main branch
- [ ] a github action that verifies the PR title (which would be used for merge commits to main)

# My research on changelogs
There are _A LOT_ of tools out there that generate changelogs from a project's git history and/or a project's git server data (like github, gitlab, and others). The [keep-a-changlog standard](https://keepachangelog.com) is a widely adopted format used for generated a changelog.

[git-cliff]: https://git-cliff.org

For the cpp_linter_rs project, I went with [`git-cliff`] because it seemed the most flexible. For other cpp-linter repos, we've been using release-drafter to keep a privately drafted release description updated until we're ready to publish. This has been ok, but it has a few problems:
1. release-drafter can be configured to follow keep-a-changelog convention, but it does not generate a changelog. See [FAQ in keep-a-changelog site about github releases](https://keepachangelog.com/en/1.1.0/#github-releases).
2. release-drafter does not account for simple commits to default branch. It only generates changes from merged PRs.
3. poor support for pre-releases. While release-drafter can be coaxed into bumping to a pre-release version, it requires manual overrides to do so. This may not be important for the entire cpp-linter org, but I noted it because cpp_linter_rs is currently in pre-release status.

## Introducing [`git-cliff`][git-cliff]

With [`git-cliff`][git-cliff], we can include direct commits main branch and merge commits from PRs.
When told the name of the github repo, [`git-cliff`][git-cliff] will fetch PR labels from GitHub's REST API.
The labels can then be used the same as release-drafter interprets them (via [cliff.toml configuration rules](https://github.com/cpp-linter/cpp_linter_rs/blob/6f7feb865ea61fb0ea6d251a58230d500839c00e/cliff.toml#L91-L131)). 

With git-cliff, we can
### generate a changelog
```
git cliff --output CHANGELOG.md
```
### generate a release description for a specific version
```
git cliff --version v2.0.0
```
### generate a release description for a the next version
Here, we can add the `--bump` option to auto-increment the version number based on unreleased changes.
```
git cliff --unreleased --bump
```
### get the next release version based on unreleased changes
```
git cliff --bumped-version
```
> [!tip]
> If we don't want to auto-bump the version number, we can manually specify the next unreleased tag instead.
> ```
> git cliff --tag 3.0.0
> ```
> Additionally, we can annotate the unreleased tag too:
> ```
> git cliff --tag 3.0.0 --with-tag-message "Some famous quote"
> ```

**Comment by shenxianpeng** [comment]
I have used [conventional commit](https://www.conventionalcommits.org/) for years, most time I can follow the spec to write commit as conventional commits, but sometimes I just want to try it out, but I still need to think of a good commit message for each commit. So, I mostly like to write a good title for the pull request instead of writing a good commit message for all the commits in the pull request to save energy and disrupt the thinking of writing code. that is way of release-drafte did. (By the way I created [conventional branch](https://conventional-branch.github.io/) recently).

There are bunch of tools but I feel they don't suite me, so I created @commit-check. 

* [commit-check](https://github.com/commit-check/commit-check) can be used as pre-commit hooks 
* [commit-check-action](https://github.com/commit-check/commit-check-action) works as a github action 

I know commit-check may not work perfect, but worth to try and appreciate to get real users's feedback to help improve it.

--- 

>release-drafter can be configured to follow keep-a-changelog convention, but it does not generate a changelog. See [FAQ in keep-a-changelog site about github releases](https://keepachangelog.com/en/1.1.0/#github-releases).

indeed. maybe for cpp_linter_rs can customize their release-drafter.yml to include a link to the changelog, such as https://cpp-linter.github.io/cpp_linter_rs/CHANGELOG.html, similar to how the Jenkins project handles it: https://github.com/jenkinsci/jenkins/releases. Personally I perfer to take a look the GitHub release tab to see the diff changes and release notes insead to go to code base to find the CHANGELOG.md.

> release-drafter does not account for simple commits to default branch. It only generates changes from merged PRs.

indeed. it shoud not friendly for projects that have just started and have a lot of changes, but for projects that have been officially released, only make change through PR has its advantages, including modified code, comments, etc., and it can avoid mistakes.

For individual project, it can be used the best way they would. For organizational projects, maintaining a consistent approach across the board is usually beneficial unless it really doesn't fit.

**Reply by 2bndy5** [reply]
The commit-check idea did cross my mind. I hadn't looked too deep into it yet though.

I have seen some people using conventional branch names. I did it once and the `/` caused some problems with `git checkout`

---
Maybe I'm just overthinking this (again).

**Reply by 2bndy5** [reply]
I'm not using release-drafter in cpp-linter-rs. That project has several package manifests that need to be updated when bumping the version number.Instead, the releases are published by [manually running a bump-n-release CI workflow](https://github.com/cpp-linter/cpp_linter_rs/actions/workflows/bump-n-release.yml). The created Github release uses the changes directly from the updated changelog (which gets pushed along with manifest changes).

**Reply by 2bndy5** [reply]
> I mostly like to write a good title for the pull request instead of writing a good commit message for all the commits

Same here. I added a free AI bot to cpp-linter org. If you create a PR titled `@coderabbitai`, then the AI will change the PR title for you (after it reviews the changes). 🤖 I also found the AI useful in finding other problems I didn't notice before.

**Reply by shenxianpeng** [reply]
`@coderabbitai` works great. I even felt a little bit like I was useless. It seems that the trial period has passed and code review is no longer supported.

**Reply by shenxianpeng** [reply]
It would be my pleasure to invite you if you are also interested in commit-check. Anyway I know you have many things to do and bandwidth may not be enough.

**Reply by 2bndy5** [reply]
It is subject to REST API rate limits. If the review fails, then there will be a message stating how long to wait for the limit to reset.

I also configured it to  not review bot-generated PRs.

**Reply by shenxianpeng** [reply]
> I have seen some people using conventional branch names. I did it once and the `/` caused some problems with `git checkout`

We've used conventional branch names integrated with Jenkins for years without encountering this issue, so it might be specific to GitHub Aciton with the `git checkout` command.

**Reply by 2bndy5** [reply]
no I mean on the command line. Using tab completion doesn't automatically add the needed quotes around branches that look like paths.

**Comment by 2bndy5** [comment]
Unreleased changes are good for project developers, but not for outsiders looking to see what changes they're using in production. Going forward I don't intend to keep the changelog updated with "Unreleased" changes. Instead the Unreleased changes can be viewed from the [bump-n-release CI's step summary](https://github.com/cpp-linter/cpp_linter_rs/actions/runs/11026487928/attempts/1#summary-30623107402).
